### PR TITLE
Annotation-based commands

### PR DESCRIPTION
**Pull request by kashike**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1201) | [**Common**](https://github.com/SpongePowered/SpongeCommon/pull/675)

### CODE REVIEW

File: src/main/java/org/spongepowered/common/command/SpongeCommandManager.java [hunk 1]
```diff
@@ -118,36 +129,64 @@ public SpongeCommandManager(Logger logger, Disambiguator disambiguator) {
...
+        checkState(module.canBeInstalledOrConfigured(false), "The provided command module has been fully configured and "
+                + "cannot be used to register commands.");
+
+        PluginContainer container = module.getPlugin();
+
+        synchronized (this.lock) {
+            List<String> aliasesWithPrefix = this.calculatePrefixesWithNamespace(container, aliases);
+
+            Optional<CommandMapping> mapping = this.dispatcher.register(callable, aliasesWithPrefix, callback,
+                    (cc, al) -> new ImmutableModuleCommandMapping(module, callable, aliases.get(0), aliases.subList(1, aliases.size())));
```
Comment (ryantheleach) [inline src/main/java/org/spongepowered/common/command/SpongeCommandManager.java]: In the API half you had something similar extracted out to a field, not sure if you wanted it done here or not. 

https://github.com/SpongePowered/SpongeAPI/pull/1201/files#diff-89adf4090b06b371ecc8b77db21c4abfR89

File: src/main/java/org/spongepowered/common/command/annotation/AnnotationCommandCallable.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents an abstract implementation of an annotation-based command.
```
Comment (ryantheleach) [inline src/main/java/org/spongepowered/common/command/annotation/AnnotationCommandCallable.java]: abstract implementation? but it's a final concrete class?

File: src/main/java/org/spongepowered/common/command/SpongeCommandManager.java (hunk 1 above)
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/SpongeCommandManager.java]: There's no such thing as a `TriFunction`, so it's created here to capture a local variable (`module`) as well.

**Comment by Feeliiix** [comment]
What's the status on this PR?

**Comment by kashike** [comment]
I'll be updating this PR as soon as I can.

**Comment by ryantheleach** [comment]
This PR has been delayed for inclusion in Dualspires Command API or afterwards.

**Comment by Zidane** [comment]
Closing this given the age that has passed and the lack of activity. If someone wants to continue the work done here then we can re-open at a later time.

Additionally this should be in the new command framework PR.

**Pull request closed without merging**